### ISSUE TITLE
Crash (double free) when doing :suspend followed by fg, introduced by patch 8.1.244

### ISSUE THREAD

**Issue by dpelle** [vim/vim#3346]
Vim-8.1.296 (huge) on Linux x86_64 sometimes crashes after
doing :suspend  (or CTRL-Z) to go in the shell, followed by the 'fg'
shell command to go back in Vim. It does not happen all the times,
but it happens frequently. Repeating :suspend  followed by fg
10 times or so is almost always enough to cause the crash, but
sometimes it crashes the first time.

I could reproduce it with asan (address sanitizer) which gave this:
```
=================================================================
==8916==ERROR: AddressSanitizer: attempting double-free on 0x6130000814c0 in thread T4:
    #0 0x7f1fabbbd2ca in __interceptor_free (/usr/lib/x86_64-linux-gnu/libasan.so.2+0x982ca)
    #1 0x671257 in vim_free /home/dope/sb/vim/src/misc2.c:1837
    #2 0x7301ac in after_sigcont /home/dope/sb/vim/src/os_unix.c:1235
    #3 0x730266 in sigcont_handler /home/dope/sb/vim/src/os_unix.c:1277
    #4 0x7f1fa98b94af  (/lib/x86_64-linux-gnu/libc.so.6+0x354af)
    #5 0x7f1fa9c5d826 in do_futex_wait.constprop.3 (/lib/x86_64-linux-gnu/libpthread.so.0+0xf826)
    #6 0x7f1fa9c5d8d3 in __new_sem_wait_slow.constprop.0 (/lib/x86_64-linux-gnu/libpthread.so.0+0xf8d3)
    #7 0x7f1fa9c5d979 in __new_sem_wait (/lib/x86_64-linux-gnu/libpthread.so.0+0xf979)
    #8 0x7f1fa9f3ffe7 in PyThread_acquire_lock (/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0+0xd4fe7)
    #9 0x7f1fa9f71831  (/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0+0x106831)
    #10 0x7f1fa9f1d970 in PyEval_EvalFrameEx (/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0+0xb2970)
    #11 0x7f1faa05305b in PyEval_EvalCodeEx (/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0+0x1e805b)
    #12 0x7f1fa9f1bf1c in PyEval_EvalFrameEx (/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0+0xb0f1c)
    #13 0x7f1faa05305b in PyEval_EvalCodeEx (/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0+0x1e805b)
    #14 0x7f1fa9f1bf1c in PyEval_EvalFrameEx (/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0+0xb0f1c)
    #15 0x7f1faa05305b in PyEval_EvalCodeEx (/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0+0x1e805b)
    #16 0x7f1fa9fa946c  (/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0+0x13e46c)
    #17 0x7f1fa9f7c272 in PyObject_Call (/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0+0x111272)
    #18 0x7f1fa9f1614b in PyEval_EvalFrameEx (/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0+0xab14b)
    #19 0x7f1fa9f1c043 in PyEval_EvalFrameEx (/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0+0xb1043)
    #20 0x7f1fa9f1c043 in PyEval_EvalFrameEx (/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0+0xb1043)
    #21 0x7f1faa05305b in PyEval_EvalCodeEx (/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0+0x1e805b)
    #22 0x7f1fa9fa936f  (/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0+0x13e36f)
    #23 0x7f1fa9f7c272 in PyObject_Call (/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0+0x111272)
    #24 0x7f1fa9ff03ab  (/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0+0x1853ab)
    #25 0x7f1fa9f7c272 in PyObject_Call (/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0+0x111272)
    #26 0x7f1faa052486 in PyEval_CallObjectWithKeywords (/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0+0x1e7486)
    #27 0x7f1fa9f750b1  (/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0+0x10a0b1)
    #28 0x7f1fa9c556b9 in start_thread (/lib/x86_64-linux-gnu/libpthread.so.0+0x76b9)
    #29 0x7f1fa998b41c in clone (/lib/x86_64-linux-gnu/libc.so.6+0x10741c)

0x6130000814c0 is located 0 bytes inside of 350-byte region [0x6130000814c0,0x61300008161e)
freed by thread T0 here:
    #0 0x7f1fabbbd2ca in __interceptor_free (/usr/lib/x86_64-linux-gnu/libasan.so.2+0x982ca)
    #1 0x671257 in vim_free /home/dope/sb/vim/src/misc2.c:1837
    #2 0x7301ac in after_sigcont /home/dope/sb/vim/src/os_unix.c:1235
    #3 0x730541 in mch_suspend /home/dope/sb/vim/src/os_unix.c:1415
    #4 0x8d2a10 in ui_suspend /home/dope/sb/vim/src/ui.c:309
    #5 0x55ff68 in ex_stop /home/dope/sb/vim/src/ex_docmd.c:7819
    #6 0x544eea in do_one_cmd /home/dope/sb/vim/src/ex_docmd.c:2503
    #7 0x53f5cf in do_cmdline /home/dope/sb/vim/src/ex_docmd.c:1041
    #8 0x53e1d2 in do_cmdline_cmd /home/dope/sb/vim/src/ex_docmd.c:642
    #9 0x6cb601 in nv_suspend /home/dope/sb/vim/src/normal.c:7903
    #10 0x6a18ee in normal_cmd /home/dope/sb/vim/src/normal.c:1134
    #11 0x9a7d41 in main_loop /home/dope/sb/vim/src/main.c:1363
    #12 0x9a7047 in vim_main2 /home/dope/sb/vim/src/main.c:919
    #13 0x9a65b8 in main /home/dope/sb/vim/src/main.c:443
    #14 0x7f1fa98a482f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)

previously allocated by thread T0 here:
    #0 0x7f1fabbbd602 in malloc (/usr/lib/x86_64-linux-gnu/libasan.so.2+0x98602)
    #1 0x66ea91 in lalloc /home/dope/sb/vim/src/misc2.c:976
    #2 0x66e8ee in alloc /home/dope/sb/vim/src/misc2.c:874
    #3 0x66f4e8 in vim_strsave /home/dope/sb/vim/src/misc2.c:1315
    #4 0x731c27 in get_x11_thing /home/dope/sb/vim/src/os_unix.c:2105
    #5 0x73176c in get_x11_title /home/dope/sb/vim/src/os_unix.c:1999
    #6 0x73208a in mch_settitle /home/dope/sb/vim/src/os_unix.c:2289
    #7 0x442ff8 in resettitle /home/dope/sb/vim/src/buffer.c:3814
    #8 0x55ff72 in ex_stop /home/dope/sb/vim/src/ex_docmd.c:7822
    #9 0x544eea in do_one_cmd /home/dope/sb/vim/src/ex_docmd.c:2503
    #10 0x53f5cf in do_cmdline /home/dope/sb/vim/src/ex_docmd.c:1041
    #11 0x53e1d2 in do_cmdline_cmd /home/dope/sb/vim/src/ex_docmd.c:642
    #12 0x6cb601 in nv_suspend /home/dope/sb/vim/src/normal.c:7903
    #13 0x6a18ee in normal_cmd /home/dope/sb/vim/src/normal.c:1134
    #14 0x9a7d41 in main_loop /home/dope/sb/vim/src/main.c:1363
    #15 0x9a7047 in vim_main2 /home/dope/sb/vim/src/main.c:919
    #16 0x9a65b8 in main /home/dope/sb/vim/src/main.c:443
    #17 0x7f1fa98a482f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)

Thread T4 created by T0 here:
    #0 0x7f1fabb5b253 in pthread_create (/usr/lib/x86_64-linux-gnu/libasan.so.2+0x36253)
    #1 0x7f1fa9f40349 in PyThread_start_new_thread (/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0+0xd5349)

SUMMARY: AddressSanitizer: double-free ??:0 __interceptor_free
==8916==ABORTING
```
It's a double free. Notice that the 2 vim_free() happen
in different threads:
* thread T0 is the main thread
* thread T4 was somehow spawned by the Python library

I use the YouCompleteMe Vim plugin that requires python
which I think is needed to trigger this bug (it may also happen
with other Vim plugins that use Python).

The bug happens in code that was recently added in this patch:
```
==============================================================================
commit 917e32bda5a93941fbbccab09ae3960114b67188
Author: Bram Moolenaar <Bram@vim.org>
Date:   Tue Aug 7 17:38:41 2018 +0200

    patch 8.1.0244: no redraw when using a STOP signal on Vim and then CONT
    
    Problem:    No redraw when using a STOP signal on Vim and then a CONT signal.
    Solution:   Catch the CONT signal and force a redraw. (closes #3285)
==============================================================================
```
I see that this patch causes to call free() inside a signal handler
which is normally not allowed. But I don't think that's the issue
here anyway. There is a race condition as after_sigcont()
is called from 2 threads. Indeed, building vim with tsan (thread
sanitizer, i.e. compiling with --fsanitize=thread), shows a race
condition:
```
==================
WARNING: ThreadSanitizer: data race (pid=11517)
  Read of size 4 at 0x000000a8b4a0 by thread T3:
    #0 sigcont_handler /home/dope/sb/vim/src/os_unix.c:1266 (vim+0x000000626f92)
    #1 <null> <null> (libc.so.6+0x0000000354af)
    #2 PyThread_acquire_lock <null> (libpython2.7.so.1.0+0x0000000d4fe7)

  Previous write of size 4 at 0x000000a8b4a0 by main thread:
    #0 mch_suspend /home/dope/sb/vim/src/os_unix.c:1380 (vim+0x000000627161)
    #1 ui_suspend /home/dope/sb/vim/src/ui.c:309 (vim+0x000000749ee4)
    #2 ex_stop /home/dope/sb/vim/src/ex_docmd.c:7819 (vim+0x0000004f67c5)
    #3 do_one_cmd /home/dope/sb/vim/src/ex_docmd.c:2503 (vim+0x0000004e51f6)
    #4 do_cmdline /home/dope/sb/vim/src/ex_docmd.c:1041 (vim+0x0000004e03f4)
    #5 nv_colon /home/dope/sb/vim/src/normal.c:5401 (vim+0x0000005db840)
    #6 normal_cmd /home/dope/sb/vim/src/normal.c:1134 (vim+0x0000005cd85a)
    #7 main_loop /home/dope/sb/vim/src/main.c:1363 (vim+0x0000007d9887)
    #8 vim_main2 /home/dope/sb/vim/src/main.c:919 (vim+0x0000007d89b0)
    #9 main /home/dope/sb/vim/src/main.c:443 (vim+0x0000007d7d3c)

  Location is global 'in_mch_suspend' of size 4 at 0x000000a8b4a0 (vim+0x000000a8b4a0)

  Thread T3 (tid=11564, running) created by main thread at:
    #0 pthread_create <null> (libtsan.so.0+0x000000027577)
    #1 PyThread_start_new_thread <null> (libpython2.7.so.1.0+0x0000000d5349)
    #2 DoPyCommand /home/dope/sb/vim/src/if_python.c:1072 (vim+0x0000007b11fa)
    #3 ex_python /home/dope/sb/vim/src/if_python.c:1116 (vim+0x0000007b1329)
    #4 do_one_cmd /home/dope/sb/vim/src/ex_docmd.c:2503 (vim+0x0000004e51f6)
    #5 do_cmdline /home/dope/sb/vim/src/ex_docmd.c:1041 (vim+0x0000004e03f4)
    #6 ex_execute /home/dope/sb/vim/src/eval.c:8177 (vim+0x00000048bad2)
    #7 do_one_cmd /home/dope/sb/vim/src/ex_docmd.c:2503 (vim+0x0000004e51f6)
    #8 do_cmdline /home/dope/sb/vim/src/ex_docmd.c:1041 (vim+0x0000004e03f4)
    #9 call_user_func /home/dope/sb/vim/src/userfunc.c:953 (vim+0x000000760793)
    #10 call_func /home/dope/sb/vim/src/userfunc.c:1458 (vim+0x0000007622bd)
    #11 get_func_tv /home/dope/sb/vim/src/userfunc.c:454 (vim+0x00000075ec56)
    #12 ex_call /home/dope/sb/vim/src/userfunc.c:3117 (vim+0x000000768cb3)
    #13 do_one_cmd /home/dope/sb/vim/src/ex_docmd.c:2503 (vim+0x0000004e51f6)
    #14 do_cmdline /home/dope/sb/vim/src/ex_docmd.c:1041 (vim+0x0000004e03f4)
    #15 apply_autocmds_group /home/dope/sb/vim/src/fileio.c:9706 (vim+0x00000053a7f5)
    #16 apply_autocmds /home/dope/sb/vim/src/fileio.c:9219 (vim+0x00000053981c)
    #17 did_set_string_option /home/dope/sb/vim/src/option.c:7717 (vim+0x0000006134d9)
    #18 set_string_option /home/dope/sb/vim/src/option.c:6021 (vim+0x00000060d81e)
    #19 set_option_value /home/dope/sb/vim/src/option.c:9873 (vim+0x000000619be7)
    #20 ex_setfiletype /home/dope/sb/vim/src/ex_docmd.c:12355 (vim+0x00000050271b)
    #21 do_one_cmd /home/dope/sb/vim/src/ex_docmd.c:2503 (vim+0x0000004e51f6)
    #22 do_cmdline /home/dope/sb/vim/src/ex_docmd.c:1041 (vim+0x0000004e03f4)
    #23 nv_colon /home/dope/sb/vim/src/normal.c:5401 (vim+0x0000005db840)
    #24 normal_cmd /home/dope/sb/vim/src/normal.c:1134 (vim+0x0000005cd85a)
    #25 main_loop /home/dope/sb/vim/src/main.c:1363 (vim+0x0000007d9887)
    #26 vim_main2 /home/dope/sb/vim/src/main.c:919 (vim+0x0000007d89b0)
    #27 main /home/dope/sb/vim/src/main.c:443 (vim+0x0000007d7d3c)

SUMMARY: ThreadSanitizer: data race /home/dope/sb/vim/src/os_unix.c:1266 sigcont_handler
==================
```

**Comment by brammool** [comment]
Dominique wrote:

> Vim-8.1.296 (huge) on Linux x86_64 sometimes crashes after
> doing :suspend  (or CTRL-Z) to go in the shell, followed by the 'fg'
> shell command to go back in Vim. It does not happen all the times,
> but it happens frequently. Repeating :suspend  followed by fg
> 10 times or so is almost always enough to cause the crash, but
> sometimes it crashes the first time.
> 
> I could reproduce it with asan (address sanitizer) which gave this:
> ```
> =================================================================
> ==8916==ERROR: AddressSanitizer: attempting double-free on 0x6130000814c0 in thread T4:
>     #0 0x7f1fabbbd2ca in __interceptor_free (/usr/lib/x86_64-linux-gnu/libasan.so.2+0x982ca)
>     #1 0x671257 in vim_free /home/dope/sb/vim/src/misc2.c:1837
>     #2 0x7301ac in after_sigcont /home/dope/sb/vim/src/os_unix.c:1235
>     #3 0x730266 in sigcont_handler /home/dope/sb/vim/src/os_unix.c:1277
>     #4 0x7f1fa98b94af  (/lib/x86_64-linux-gnu/libc.so.6+0x354af)

[...]

> 
> 0x6130000814c0 is located 0 bytes inside of 350-byte region [0x6130000814c0,0x61300008161e)
> freed by thread T0 here:
>     #0 0x7f1fabbbd2ca in __interceptor_free (/usr/lib/x86_64-linux-gnu/libasan.so.2+0x982ca)
>     #1 0x671257 in vim_free /home/dope/sb/vim/src/misc2.c:1837
>     #2 0x7301ac in after_sigcont /home/dope/sb/vim/src/os_unix.c:1235
>     #3 0x730541 in mch_suspend /home/dope/sb/vim/src/os_unix.c:1415

[...]

> previously allocated by thread T0 here:
>     #0 0x7f1fabbbd602 in malloc (/usr/lib/x86_64-linux-gnu/libasan.so.2+0x98602)
>     #1 0x66ea91 in lalloc /home/dope/sb/vim/src/misc2.c:976
>     #2 0x66e8ee in alloc /home/dope/sb/vim/src/misc2.c:874
>     #3 0x66f4e8 in vim_strsave /home/dope/sb/vim/src/misc2.c:1315
>     #4 0x731c27 in get_x11_thing /home/dope/sb/vim/src/os_unix.c:2105
>     #5 0x73176c in get_x11_title /home/dope/sb/vim/src/os_unix.c:1999
>     #6 0x73208a in mch_settitle /home/dope/sb/vim/src/os_unix.c:2289

[...]

> It's a double free. Notice that the 2 vim_free() happen
> in different threads:
> * thread T0 is the main thread
> * thread T4 was somehow spawned by the Python library
> 
> I use the YouCompleteMe Vim plugin that requires python
> which I think is needed to trigger this bug (it may also happen
> with other Vim plugins that use Python).
> 
> The bug happens in code that was recently added in this patch:
> ```
> ==============================================================================
> commit 917e32bda5a93941fbbccab09ae3960114b67188
> Author: Bram Moolenaar <Bram@vim.org>
> Date:   Tue Aug 7 17:38:41 2018 +0200
> 
>     patch 8.1.0244: no redraw when using a STOP signal on Vim and then CONT
>     
>     Problem:    No redraw when using a STOP signal on Vim and then a CONT signal.
>     Solution:   Catch the CONT signal and force a redraw. (closes #3285)
> ==============================================================================
> ```
> I see that this patch causes to call free() inside a signal handler
> which is normally not allowed. But I don't think that's the issue
> here anyway. There is a race condition as after_sigcont()
> is called from 2 threads. Indeed, building vim with tsan (thread
> sanitizer, i.e. compiling with --fsanitize=thread), shows a race
> condition:

We probably need to make variables used in signal functions "volatile".

Another thing is that freeing memory in a signal handler might not be
safe.  It might be OK for SIGCONT, since we handle it only when
suspended.

Can you try what happens if you make "oldtitle" volatile?

Note that in sigcont_handler() we do much more that might normally not
be OK in a signal handler, but since it's done in suspended state it
might be OK.

Hmm, maybe "in_mch_suspend" also needs to be volatile?

-- 
People who want to share their religious views with you
almost never want you to share yours with them.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by dpelle** [comment]
@brammool  wrote:

> We probably need to make variables used in signal functions "volatile". 

No. That won't fix the race condition. In fact, "volatile" is never the
answer for race conditions with multiple threads.

One problem at least, is that VIM_CLEAR called in after_sigcont()
is not atomic:
```
1230      static void
1231  after_sigcont(void)
1232  {
1233  # ifdef FEAT_TITLE
1234      // Set oldtitle to NULL, so the current title is obtained again.
1235      VIM_CLEAR(oldtitle);
1236  # endif
 ````

With VIM_CLEAR is defined in macro.h as:
```
370  #define VIM_CLEAR(p) \
371      do { \
372  	if ((p) != NULL) { \
373  	    vim_free(p); \
374  	    (p) = NULL; \
375  	} \
376      } while (0)
```
The 2 threads calling VIM_CLEAR(oldtitle) may call 
vim_free(oldtile) at line macro.h:373 at the same time
before oldtitle is set to NULL (hence double free) at line
macro.h:374. That's what happens according the stack
given by asan.
 
We can't take a mutex either because it's in a signal
handler.

As you wrote, we should not be doing so much in the
signal handler either. Calling free() is not safe in a signal
handler, but that's a 2nd issue.

I don't really know why libpython is spawning
a thread and whether that thread really needs to
handle sigcont_handler(). Maybe it's the YouCompleteMe
plugin which is spawning a thread. But YouCompleteMe
was working fine until patch 8.1.244.

It's quite a severe issue, as the crash has happened
multiple times to me while using Vim normally (i.e. without
doing odd things). I suspect it will affect many other users
of YouCompleteMe.

**Comment by brammool** [comment]
Dominique wrote:

> @brammool  wrote:
> 
> > We probably need to make variables used in signal functions "volatile". 
> 
> No. That won't fix the race condition. In fact, "volatile" is never the
> answer for race conditions with multiple threads.
> 
> One problem at least, is that VIM_CLEAR called in after_sigcont()
> is not atomic:
> ```
> 1230      static void
> 1231  after_sigcont(void)
> 1232  {
> 1233  # ifdef FEAT_TITLE
> 1234      // Set oldtitle to NULL, so the current title is obtained again.
> 1235      VIM_CLEAR(oldtitle);
> 1236  # endif
>  ````
> 
> With VIM_CLEAR is defined in macro.h as:
> ```
> 370  #define VIM_CLEAR(p) \
> 371      do { \
> 372  	if ((p) != NULL) { \
> 373  	    vim_free(p); \
> 374  	    (p) = NULL; \
> 375  	} \
> 376      } while (0)
> ```
> The 2 threads calling VIM_CLEAR(oldtitle) may call 
> vim_free(oldtile) at line macro.h:373 at the same time
> before oldtitle is set to NULL (hence double free) at line
> macro.h:374. That's what happens according the stack
> given by asan.
>  
> We can't take a mutex either because it's in a signal
> handler.
> 
> As you wrote, we should not be doing so much in the
> signal handler either. Calling free() is not safe in a signal
> handler, but that's a 2nd issue.
> 
> I don't really know why libpython is spawning
> a thread and whether that thread really needs to
> handle sigcont_handler(). Maybe it's the YouCompleteMe
> plugin which is spawning a thread. But YouCompleteMe
> was working fine until patch 8.1.244.
> 
> It's quite a severe issue, as the crash has happened
> multiple times to me while using Vim normally (i.e. without
> doing odd things). I suspect it will affect many other users
> of YouCompleteMe.

I cannot reproduce it, but I can guess a solution.  I'll make a patch,
please try it out.

-- 
It might look like I'm doing nothing, but at the cellular level
I'm really quite busy.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by dpelle** [comment]
@brammool wrote:

> I cannot reproduce it, but I can guess a solution. I'll make a patch, please try it out.

I tried to reproduce it with vim-8.1.301.
Unfortunately, it still crashes once in a while when doing ":suspend" 
followed by fg to get back to vim.  Sometimes, it does not crash
immediately but Vim screen gets garbled. Sometimes it crashes
immediately.  I see errors with both asan and tsan.  With asan, 
the reported stacks are not always the same and tsan reports
race conditions, which might explain why behavior is not always
the same.

Here are 2 slightly different asan outputs when reproducing the bug
(with suspend followed by fg in shell):

*Crash 1*:
```
=================================================================
==4229==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x60200000a1d6 at pc 0x000000775852 bp 0x7fffcedfeac0 sp 0x7fffcedfeab0
READ of size 1 at 0x60200000a1d6 thread T0
    #0 0x775851 in bt_regexec_both /home/dope/sb/vim/src/regexp.c:3974
    #1 0x774b7f in bt_regexec_nl /home/dope/sb/vim/src/regexp.c:3740
    #2 0x7b3c2d in vim_regexec_string /home/dope/sb/vim/src/regexp.c:8268
    #3 0x7b41ae in vim_regexec_prog /home/dope/sb/vim/src/regexp.c:8322
    #4 0x7310e6 in x_connect_to_server /home/dope/sb/vim/src/os_unix.c:1795
    #5 0x73136a in get_x11_windis /home/dope/sb/vim/src/os_unix.c:1888
    #6 0x731fd3 in mch_settitle /home/dope/sb/vim/src/os_unix.c:2271
    #7 0x442ff8 in resettitle /home/dope/sb/vim/src/buffer.c:3814
    #8 0x55ff72 in ex_stop /home/dope/sb/vim/src/ex_docmd.c:7822
    #9 0x544eea in do_one_cmd /home/dope/sb/vim/src/ex_docmd.c:2503
    #10 0x53f5cf in do_cmdline /home/dope/sb/vim/src/ex_docmd.c:1041
    #11 0x53e1d2 in do_cmdline_cmd /home/dope/sb/vim/src/ex_docmd.c:642
    #12 0x6cb601 in nv_suspend /home/dope/sb/vim/src/normal.c:7903
    #13 0x6a18ee in normal_cmd /home/dope/sb/vim/src/normal.c:1134
    #14 0x9a7d5f in main_loop /home/dope/sb/vim/src/main.c:1363
    #15 0x9a7065 in vim_main2 /home/dope/sb/vim/src/main.c:919
    #16 0x9a65d6 in main /home/dope/sb/vim/src/main.c:443
    #17 0x7fb78c41f82f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)
    #18 0x42f198 in _start (/home/dope/sb/vim/src/vim+0x42f198)

0x60200000a1d6 is located 0 bytes to the right of 6-byte region [0x60200000a1d0,0x60200000a1d6)
allocated by thread T0 here:
    #0 0x7fb78e738602 in malloc (/usr/lib/x86_64-linux-gnu/libasan.so.2+0x98602)
    #1 0x66ea91 in lalloc /home/dope/sb/vim/src/misc2.c:976
    #2 0x66e8ee in alloc /home/dope/sb/vim/src/misc2.c:874
    #3 0x66f4e8 in vim_strsave /home/dope/sb/vim/src/misc2.c:1315
    #4 0x708a83 in set_string_option_direct /home/dope/sb/vim/src/option.c:5922
    #5 0x8a8031 in termcapinit /home/dope/sb/vim/src/term.c:2510
    #6 0x9a6490 in main /home/dope/sb/vim/src/main.c:383
    #7 0x7fb78c41f82f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)

SUMMARY: AddressSanitizer: heap-buffer-overflow /home/dope/sb/vim/src/regexp.c:3974 bt_regexec_both
Shadow bytes around the buggy address:
  0x0c047fff93e0: fa fa 00 fa fa fa 00 fa fa fa 00 fa fa fa 00 fa
  0x0c047fff93f0: fa fa 00 fa fa fa 00 fa fa fa 00 fa fa fa 00 fa
  0x0c047fff9400: fa fa 05 fa fa fa 05 fa fa fa 05 fa fa fa 05 fa
  0x0c047fff9410: fa fa 07 fa fa fa 07 fa fa fa 07 fa fa fa 07 fa
  0x0c047fff9420: fa fa 05 fa fa fa 05 fa fa fa 05 fa fa fa 05 fa
=>0x0c047fff9430: fa fa 06 fa fa fa 06 fa fa fa[06]fa fa fa 07 fa
  0x0c047fff9440: fa fa 00 06 fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c047fff9450: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c047fff9460: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c047fff9470: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c047fff9480: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Heap right redzone:      fb
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack partial redzone:   f4
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
==4229==ABORTING
```

*Crash 2*:
```
=================================================================
==5280==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x60200000a1d6 at pc 0x000000694f8d bp 0x7ffc7d5df5e0 sp 0x7ffc7d5df5d0
READ of size 1 at 0x60200000a1d6 thread T0
    #0 0x694f8c in utf_ptr2char /home/dope/sb/vim/src/mbyte.c:1795
    #1 0x7aad70 in nfa_regmatch /home/dope/sb/vim/src/regexp_nfa.c:5600
    #2 0x7b1a13 in nfa_regtry /home/dope/sb/vim/src/regexp_nfa.c:6972
    #3 0x7b2db3 in nfa_regexec_both /home/dope/sb/vim/src/regexp_nfa.c:7172
    #4 0x7b353a in nfa_regexec_nl /home/dope/sb/vim/src/regexp_nfa.c:7339
    #5 0x7b3c2d in vim_regexec_string /home/dope/sb/vim/src/regexp.c:8268
    #6 0x7b41ae in vim_regexec_prog /home/dope/sb/vim/src/regexp.c:8322
    #7 0x7310e6 in x_connect_to_server /home/dope/sb/vim/src/os_unix.c:1795
    #8 0x73136a in get_x11_windis /home/dope/sb/vim/src/os_unix.c:1888
    #9 0x731fd3 in mch_settitle /home/dope/sb/vim/src/os_unix.c:2271
    #10 0x442ff8 in resettitle /home/dope/sb/vim/src/buffer.c:3814
    #11 0x55ff72 in ex_stop /home/dope/sb/vim/src/ex_docmd.c:7822
    #12 0x544eea in do_one_cmd /home/dope/sb/vim/src/ex_docmd.c:2503
    #13 0x53f5cf in do_cmdline /home/dope/sb/vim/src/ex_docmd.c:1041
    #14 0x53e1d2 in do_cmdline_cmd /home/dope/sb/vim/src/ex_docmd.c:642
    #15 0x6cb601 in nv_suspend /home/dope/sb/vim/src/normal.c:7903
    #16 0x6a18ee in normal_cmd /home/dope/sb/vim/src/normal.c:1134
    #17 0x9a7d5f in main_loop /home/dope/sb/vim/src/main.c:1363
    #18 0x9a7065 in vim_main2 /home/dope/sb/vim/src/main.c:919
    #19 0x9a65d6 in main /home/dope/sb/vim/src/main.c:443
    #20 0x7f6521f4882f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)
    #21 0x42f198 in _start (/home/dope/sb/vim/src/vim+0x42f198)

0x60200000a1d6 is located 0 bytes to the right of 6-byte region [0x60200000a1d0,0x60200000a1d6)
allocated by thread T0 here:
    #0 0x7f6524261602 in malloc (/usr/lib/x86_64-linux-gnu/libasan.so.2+0x98602)
    #1 0x66ea91 in lalloc /home/dope/sb/vim/src/misc2.c:976
    #2 0x66e8ee in alloc /home/dope/sb/vim/src/misc2.c:874
    #3 0x66f4e8 in vim_strsave /home/dope/sb/vim/src/misc2.c:1315
    #4 0x708a83 in set_string_option_direct /home/dope/sb/vim/src/option.c:5922
    #5 0x8a8031 in termcapinit /home/dope/sb/vim/src/term.c:2510
    #6 0x9a6490 in main /home/dope/sb/vim/src/main.c:383
    #7 0x7f6521f4882f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)

SUMMARY: AddressSanitizer: heap-buffer-overflow /home/dope/sb/vim/src/mbyte.c:1795 utf_ptr2char
Shadow bytes around the buggy address:
  0x0c047fff93e0: fa fa 00 fa fa fa 00 fa fa fa 00 fa fa fa 00 fa
  0x0c047fff93f0: fa fa 00 fa fa fa 00 fa fa fa 00 fa fa fa 00 fa
  0x0c047fff9400: fa fa 05 fa fa fa 05 fa fa fa 05 fa fa fa 05 fa
  0x0c047fff9410: fa fa 07 fa fa fa 07 fa fa fa 07 fa fa fa 07 fa
  0x0c047fff9420: fa fa 05 fa fa fa 05 fa fa fa 05 fa fa fa 05 fa
=>0x0c047fff9430: fa fa 06 fa fa fa 06 fa fa fa[06]fa fa fa 07 fa
  0x0c047fff9440: fa fa 00 06 fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c047fff9450: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c047fff9460: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c047fff9470: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c047fff9480: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Heap right redzone:      fb
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack partial redzone:   f4
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
==5280==ABORTING
```

The tsan (thread sanitizer) output is quite verbose, but here is the
first tsan error:

```
==================
WARNING: ThreadSanitizer: data race (pid=3399)
  Read of size 4 at 0x000000a8b4a0 by thread T2:
    #0 sigcont_handler /home/dope/sb/vim/src/os_unix.c:1268 (vim+0x000000626f62)
    #1 <null> <null> (libc.so.6+0x0000000354af)
    #2 PyThread_acquire_lock <null> (libpython2.7.so.1.0+0x0000000d4fe7)

  Previous write of size 4 at 0x000000a8b4a0 by main thread:
    #0 mch_suspend /home/dope/sb/vim/src/os_unix.c:1382 (vim+0x000000627131)
    #1 ui_suspend /home/dope/sb/vim/src/ui.c:309 (vim+0x000000749f28)
    #2 ex_stop /home/dope/sb/vim/src/ex_docmd.c:7819 (vim+0x0000004f67c5)
    #3 do_one_cmd /home/dope/sb/vim/src/ex_docmd.c:2503 (vim+0x0000004e51f6)
    #4 do_cmdline /home/dope/sb/vim/src/ex_docmd.c:1041 (vim+0x0000004e03f4)
    #5 do_cmdline_cmd /home/dope/sb/vim/src/ex_docmd.c:642 (vim+0x0000004df501)
    #6 nv_suspend /home/dope/sb/vim/src/normal.c:7903 (vim+0x0000005e4dd7)
    #7 normal_cmd /home/dope/sb/vim/src/normal.c:1134 (vim+0x0000005cd85a)
    #8 main_loop /home/dope/sb/vim/src/main.c:1363 (vim+0x0000007d98cb)
    #9 vim_main2 /home/dope/sb/vim/src/main.c:919 (vim+0x0000007d89f4)
    #10 main /home/dope/sb/vim/src/main.c:443 (vim+0x0000007d7d80)

  Location is global 'in_mch_suspend' of size 4 at 0x000000a8b4a0 (vim+0x000000a8b4a0)

  Thread T2 (tid=3442, running) created by main thread at:
    #0 pthread_create <null> (libtsan.so.0+0x000000027577)
    #1 PyThread_start_new_thread <null> (libpython2.7.so.1.0+0x0000000d5349)
    #2 DoPyCommand /home/dope/sb/vim/src/if_python.c:1072 (vim+0x0000007b123e)
    #3 do_pyeval /home/dope/sb/vim/src/if_python.c:1558 (vim+0x0000007b221c)
    #4 f_pyeval /home/dope/sb/vim/src/evalfunc.c:8731 (vim+0x0000004a7393)
    #5 call_internal_func /home/dope/sb/vim/src/evalfunc.c:1068 (vim+0x0000004924b8)
    #6 call_func /home/dope/sb/vim/src/userfunc.c:1477 (vim+0x0000007623f8)
    #7 get_func_tv /home/dope/sb/vim/src/userfunc.c:454 (vim+0x00000075ec9a)
    #8 eval7 /home/dope/sb/vim/src/eval.c:4142 (vim+0x00000047fb7d)
    #9 eval6 /home/dope/sb/vim/src/eval.c:3779 (vim+0x00000047ee1c)
    #10 eval5 /home/dope/sb/vim/src/eval.c:3595 (vim+0x00000047e701)
    #11 eval4 /home/dope/sb/vim/src/eval.c:3479 (vim+0x00000047e312)
    #12 eval3 /home/dope/sb/vim/src/eval.c:3399 (vim+0x00000047e0ca)
    #13 eval2 /home/dope/sb/vim/src/eval.c:3331 (vim+0x00000047de84)
    #14 eval1 /home/dope/sb/vim/src/eval.c:3259 (vim+0x00000047dbeb)
    #15 eval0 /home/dope/sb/vim/src/eval.c:3219 (vim+0x00000047dadc)
    #16 ex_return /home/dope/sb/vim/src/userfunc.c:2997 (vim+0x000000768654)
    #17 do_one_cmd /home/dope/sb/vim/src/ex_docmd.c:2503 (vim+0x0000004e51f6)
    #18 do_cmdline /home/dope/sb/vim/src/ex_docmd.c:1041 (vim+0x0000004e03f4)
    #19 call_user_func /home/dope/sb/vim/src/userfunc.c:953 (vim+0x0000007607d7)
    #20 call_func /home/dope/sb/vim/src/userfunc.c:1458 (vim+0x000000762301)
    #21 get_func_tv /home/dope/sb/vim/src/userfunc.c:454 (vim+0x00000075ec9a)
    #22 eval7 /home/dope/sb/vim/src/eval.c:4142 (vim+0x00000047fb7d)
    #23 eval6 /home/dope/sb/vim/src/eval.c:3779 (vim+0x00000047ee1c)
    #24 eval5 /home/dope/sb/vim/src/eval.c:3595 (vim+0x00000047e701)
    #25 eval4 /home/dope/sb/vim/src/eval.c:3479 (vim+0x00000047e312)
    #26 eval3 /home/dope/sb/vim/src/eval.c:3399 (vim+0x00000047e0ca)
    #27 eval2 /home/dope/sb/vim/src/eval.c:3331 (vim+0x00000047de84)
    #28 eval1 /home/dope/sb/vim/src/eval.c:3259 (vim+0x00000047dbeb)
    #29 eval0 /home/dope/sb/vim/src/eval.c:3219 (vim+0x00000047dadc)
    #30 eval_to_bool /home/dope/sb/vim/src/eval.c:678 (vim+0x000000475141)
    #31 ex_if /home/dope/sb/vim/src/ex_eval.c:904 (vim+0x000000505160)
    #32 do_one_cmd /home/dope/sb/vim/src/ex_docmd.c:2503 (vim+0x0000004e51f6)
    #33 do_cmdline /home/dope/sb/vim/src/ex_docmd.c:1041 (vim+0x0000004e03f4)
    #34 call_user_func /home/dope/sb/vim/src/userfunc.c:953 (vim+0x0000007607d7)
    #35 call_func /home/dope/sb/vim/src/userfunc.c:1458 (vim+0x000000762301)
    #36 timer_callback /home/dope/sb/vim/src/ex_cmds2.c:1302 (vim+0x0000004d2c5a)
    #37 check_due_timer /home/dope/sb/vim/src/ex_cmds2.c:1366 (vim+0x0000004d2f93)
    #38 ui_wait_for_chars_or_timer /home/dope/sb/vim/src/ui.c:234 (vim+0x000000749df6)
    #39 WaitForChar /home/dope/sb/vim/src/os_unix.c:5985 (vim+0x00000062f7e7)
    #40 mch_inchar /home/dope/sb/vim/src/os_unix.c:488 (vim+0x0000006264a1)
    #41 ui_inchar /home/dope/sb/vim/src/ui.c:190 (vim+0x000000749d1f)
    #42 inchar /home/dope/sb/vim/src/getchar.c:3099 (vim+0x00000054ff87)
    #43 vgetorpeek /home/dope/sb/vim/src/getchar.c:2881 (vim+0x00000054f9a0)
    #44 vgetc /home/dope/sb/vim/src/getchar.c:1634 (vim+0x00000054c975)
    #45 safe_vgetc /home/dope/sb/vim/src/getchar.c:1838 (vim+0x00000054cfd9)
    #46 normal_cmd /home/dope/sb/vim/src/normal.c:621 (vim+0x0000005cb978)
    #47 main_loop /home/dope/sb/vim/src/main.c:1363 (vim+0x0000007d98cb)
    #48 vim_main2 /home/dope/sb/vim/src/main.c:919 (vim+0x0000007d89f4)
    #49 main /home/dope/sb/vim/src/main.c:443 (vim+0x0000007d7d80)

SUMMARY: ThreadSanitizer: data race /home/dope/sb/vim/src/os_unix.c:1268 sigcont_handler
==================
```
So according to tsan, the in_mch_suspend variable is accessed from 2
different threads and it is not atomic nor protected by a mutex (hence
race condition).  It cannot be protected by a mutex since it's used in
signal handler.  Atomic variable are only availble in C11 I believe.

The fact that that there is another thread than the main thread seems to
 be caused by the YouCompleteMe plugin that may spawn a thread, which 
may not be safe given that Vim code is not thread safe in general. So
perhaps @micbou or @Valloric have ideas about this bug.

I'll try to minimize my setup to reproduce it in a easier way.  But in a
nutshell, reproducing it requires the YouCompleteMe plugin, then
opening a file in Vim and doing  CTRL-Z (or ":suspend")  followed
by the fg command in shell to get back to vim  (crash then sometimes
happens). Crash happens with a without a sanitized build, but it's
best to try to reproduce with a asan build (more chances to crashes
and it gives better information when bug happens).

**Comment by desdic** [comment]
I also get coredumps by pressing CTRL+z og resuming .. not every time but maybe every 10th time.

➜  ~ vim --version
VIM - Vi IMproved 8.1 (2018 May 18, compiled Aug 13 2018 14:59:01)

gdb-peda$ info threads 
  Id   Target Id         Frame 
* 1    Thread 0x7fbedda01700 (LWP 20202) 0x00007fbee161eb42 in __strcmp_avx2 () from /usr/lib/libc.so.6
  2    Thread 0x7fbee12bc740 (LWP 20154) 0x00007fbee161eb42 in __strcmp_avx2 () from /usr/lib/libc.so.6
  3    Thread 0x7fbedc9ff700 (LWP 20204) 0x00007fbee169b406 in recv () from /usr/lib/libpthread.so.0
  4    Thread 0x7fbecf7fe700 (LWP 20244) 0x00007fbee169b406 in recv () from /usr/lib/libpthread.so.0
  5    Thread 0x7fbece7fc700 (LWP 20248) 0x00007fbee169b406 in recv () from /usr/lib/libpthread.so.0
  6    Thread 0x7fbecffff700 (LWP 20238) 0x00007fbee169b406 in recv () from /usr/lib/libpthread.so.0
  7    Thread 0x7fbecdffb700 (LWP 20249) 0x00007fbee169b406 in recv () from /usr/lib/libpthread.so.0
  8    Thread 0x7fbedd200700 (LWP 20203) 0x00007fbee15b4bb1 in poll () from /usr/lib/libc.so.6
  9    Thread 0x7fbeceffd700 (LWP 20245) 0x00007fbee169b406 in recv () from /usr/lib/libpthread.so.0
gdb-peda$ bt
#0  0x00007fbee161eb42 in __strcmp_avx2 () from /usr/lib/libc.so.6
#1  0x000055b4ee765b00 in handle_viminfo_history ()
#2  0x000055b4ee739915 in ?? ()
#3  0x000055b4ee739e35 in write_viminfo ()
#4  0x000055b4ee8da0c1 in getout ()
#5  <signal handler called>
#6  0x000055b4ee7feaef in ?? ()
#7  0x000055b4ee818342 in ?? ()
#8  0x000055b4ee818772 in ?? ()
#9  0x000055b4ee81d60a in vim_regexec_multi ()
#10 0x000055b4ee8586a4 in ?? ()
#11 0x000055b4ee861605 in ?? ()
#12 0x000055b4ee862ed0 in get_syntax_attr ()
#13 0x000055b4ee8294aa in ?? ()
#14 0x000055b4ee82d7c8 in ?? ()
#15 0x000055b4ee83028a in update_screen ()
#16 0x000055b4ee7eac14 in ?? ()
#17 <signal handler called>
#18 0x00007fbee15b4bb1 in poll () from /usr/lib/libc.so.6
#19 0x00007fbedf6f5cbf in ?? () from /usr/lib/python3.7/lib-dynload/select.cpython-37m-x86_64-linux-gnu.so
#20 0x00007fbee0839f68 in _PyMethodDef_RawFastCallKeywords () from /usr/lib/libpython3.7m.so.1.0
#21 0x00007fbee0866c6f in _PyMethodDescr_FastCallKeywords () from /usr/lib/libpython3.7m.so.1.0
#22 0x00007fbee08afaee in _PyEval_EvalFrameDefault () from /usr/lib/libpython3.7m.so.1.0
#23 0x00007fbee07f2069 in _PyEval_EvalCodeWithName () from /usr/lib/libpython3.7m.so.1.0
#24 0x00007fbee0839982 in _PyFunction_FastCallKeywords () from /usr/lib/libpython3.7m.so.1.0
#25 0x00007fbee08ab225 in _PyEval_EvalFrameDefault () from /usr/lib/libpython3.7m.so.1.0
#26 0x00007fbee07f2069 in _PyEval_EvalCodeWithName () from /usr/lib/libpython3.7m.so.1.0
#27 0x00007fbee07f3268 in _PyFunction_FastCallDict () from /usr/lib/libpython3.7m.so.1.0
#28 0x00007fbee08024f8 in _PyObject_Call_Prepend () from /usr/lib/libpython3.7m.so.1.0
#29 0x00007fbee07f37cb in PyObject_Call () from /usr/lib/libpython3.7m.so.1.0
#30 0x00007fbee08aca76 in _PyEval_EvalFrameDefault () from /usr/lib/libpython3.7m.so.1.0
#31 0x00007fbee08397db in _PyFunction_FastCallKeywords () from /usr/lib/libpython3.7m.so.1.0
#32 0x00007fbee08ab225 in _PyEval_EvalFrameDefault () from /usr/lib/libpython3.7m.so.1.0
#33 0x00007fbee08397db in _PyFunction_FastCallKeywords () from /usr/lib/libpython3.7m.so.1.0
#34 0x00007fbee08ab225 in _PyEval_EvalFrameDefault () from /usr/lib/libpython3.7m.so.1.0
#35 0x00007fbee07f308b in _PyFunction_FastCallDict () from /usr/lib/libpython3.7m.so.1.0
#36 0x00007fbee08024f8 in _PyObject_Call_Prepend () from /usr/lib/libpython3.7m.so.1.0
#37 0x00007fbee07f37cb in PyObject_Call () from /usr/lib/libpython3.7m.so.1.0
#38 0x00007fbee0925c77 in ?? () from /usr/lib/libpython3.7m.so.1.0
#39 0x00007fbee1691a9d in start_thread () from /usr/lib/libpthread.so.0
#40 0x00007fbee15bfa43 in clone () from /usr/lib/libc.so.6

**Comment by dpelle** [comment]
@desdic wrote:

> I also get coredumps by pressing CTRL+z og resuming .. not every time but maybe every 10th time.

I assume you meant fg  (not og).

Do you also use the YouCompleteMe plugin like me? or else, which other plugin
do you use that triggers the bug?

Your vim does not seem to have symbols, so the stack is not very useful.
Best to comment out the lines that strips vim in vim/src/Makefile when reporting
a stack i.e. uncomment out this line:

STRIP = /bin/true

On my side, I verified that in the older vim-8.1.243, tsan (thread sanitizer)
and asan (address sanitizer) gave no errors, and no crashes. Troubles happen
from vim-8.1.244.

**Comment by desdic** [comment]
ah sorry yes I ment fg.. no I don't use YouCompleteMe so it might be a different bug. I do however use deoplete .. just installed via pacman but I will try to recompile to reproduce

**Comment by micbou** [comment]
I am trying to reproduce the issue but compiling with TSAN enabled (`-fsanitizer=thread`) always causes Vim to terminate with a bunch of warnings
```
WARNING: ThreadSanitizer: signal-unsafe call inside of a signal (pid=...)
```
after running a Python command (e.g. `:py3 pass`). Am I missing something?

> The fact that that there is another thread than the main thread seems to
be caused by the YouCompleteMe plugin that may spawn a thread, which
may not be safe given that Vim code is not thread safe in general. So
perhaps @micbou or @Valloric have ideas about this bug.

The threads spawned by YCM only communicate with [the ycmd server](https://github.com/Valloric/ycmd). They shouldn't execute Vim code.

**Comment by dpelle** [comment]
Here is a small ~/.vimrc to reproduce it.  It assumes that the Vundle and YouCompleteMe are installed. I have the latest versions of those plugins as of today.
```
$ cat ~/.vimrc
filetype off
set rtp+=~/.vim/bundle/Vundle.vim
call vundle#begin()
Plugin 'VundleVim/Vundle.vim'
Plugin 'Valloric/YouCompleteMe'
call vundle#end()
filetype plugin indent on
syntax on
```

I use this small change to make sure that Vim is built with asan:
```
$ git diff
diff --git a/src/Makefile b/src/Makefile
index f2fafa4..717cadb 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -690,7 +690,7 @@ LINT_OPTIONS = -beprxzF
 # address sanitizer or with the undefined sanitizer.  Works with gcc and
 # clang.  May make Vim twice as slow.  Errors reported on stderr.
 # More at: https://code.google.com/p/address-sanitizer/
-#SANITIZER_CFLAGS = -g -O0 -fsanitize=address -fno-omit-frame-pointer
+SANITIZER_CFLAGS = -g -O0 -fsanitize=address -fno-omit-frame-pointer
 #SANITIZER_CFLAGS = -g -O0 -fsanitize=undefined -fno-omit-frame-pointer
 SANITIZER_LIBS = $(SANITIZER_CFLAGS)
 
@@ -1156,7 +1156,7 @@ INSTALL_DATA_R    = cp -r
 
 ### Program to run on installed binary.  Use the second one to disable strip.
 #STRIP = strip
-#STRIP = /bin/true
+STRIP = /bin/true
 
 ### Permissions for binaries  {{{1
 BINMOD = 755
```

And I build Vim 8.1.301 (latest) as follows (python interpreter is needed for YouCompleteMe):
```
$ cd vim/src
$ ./configure --with-features=huge --enable-gui=none --enable-pythoninterp=yes
$ make clean
$ make -j8
```

Then I do:
```
$ cd vim/src
$ ./vim main.c 2> asan.log
```

I then press CTRL-z  to go in the shell, followed by the fg shell builtin command to go back to vim. I do that several times until vim eventually crashes. Sometimes doing it once is enough, sometimes I need 10 times or so before it crashes, but it always crashes eventually.

It does not always crash in the same way.  The last time I tried, asan did not report anything but it crashed with this output:
```
[xcb] Unknown request in queue while dequeuing
[xcb] Most likely this is a multi-threaded client and XInitThreads has not been called
[xcb] Aborting, sorry about that.
vim: ../../src/xcb_io.c:179: dequeue_pending_request: Assertion `!xcb_xlib_unknown_req_in_deq' failed.
```
But more often, asan detects errors as reported earlier.

Now If I built Vim with tsan instead of asan with this patch:
```
$ git diff
diff --git a/src/Makefile b/src/Makefile
index f2fafa4..717cadb 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -690,7 +690,7 @@ LINT_OPTIONS = -beprxzF
 # address sanitizer or with the undefined sanitizer.  Works with gcc and
 # clang.  May make Vim twice as slow.  Errors reported on stderr.
 # More at: https://code.google.com/p/address-sanitizer/
-#SANITIZER_CFLAGS = -g -O0 -fsanitize=address -fno-omit-frame-pointer
+SANITIZER_CFLAGS = -g -O0 -fsanitize=thread -fno-omit-frame-pointer
 #SANITIZER_CFLAGS = -g -O0 -fsanitize=undefined -fno-omit-frame-pointer
 SANITIZER_LIBS = $(SANITIZER_CFLAGS)
 
@@ -1156,7 +1156,7 @@ INSTALL_DATA_R    = cp -r
 
 ### Program to run on installed binary.  Use the second one to disable strip.
 #STRIP = strip
-#STRIP = /bin/true
+STRIP = /bin/true
 
 ### Permissions for binaries  {{{1
 BINMOD = 755
```

And run "./vim main.c 2> tsan.log", then type CTRL-Z  followed by 'fg' I
immediately always get many errors in tsan.log as soon as I type fg, even
if it does not always crash. SInce tsan output is quote verbose, I will attach its
output  "tsan.log":

[tsan.log](https://github.com/vim/vim/files/2302856/tsan.log)

I'm using xubuntu-16.04 (x86_64), up-to-date, in case it matters.

@micbou wrote:

> The threads spawned by YCM only communicate with the ycmd server. They shouldn't execute Vim code.

But the YCM thread I suppose handle signals, which run Vim function sigcont_handler()
which is not thread safe.

**Comment by micbou** [comment]
Thanks for the detailed instructions. Vim now doesn't exit immediately. However, I can't reproduce your issue. There must be something else.

> But the YCM thread I suppose handle signals, which run Vim function sigcont_handler()
which is not thread safe.

According to the Python docs, [only the main thread accepts signals](https://docs.python.org/2/library/signal.html):

> only the main thread can set a new signal handler, and the main thread will be the only one to receive signals (this is enforced by the Python signal module, even if the underlying thread implementation supports sending signals to individual threads).

**Comment by dpelle** [comment]
@micbou wrote:

> [...] compiling with TSAN enabled (-fsanitizer=thread) always causes Vim to
> terminate with a bunch of warnings [...] after running a Python command
> (e.g. :py3 pass). Am I missing something?

I do not see that, on xubuntu-16.04.  For me, tsan reports no error when
simply running ":py pass" or ":py3 pass", with vim started with "vim --clean"
and obviously vim built with python interpreter or python3 interpreter.

**Comment by micbou** [comment]
@dpelle Does Vim still crash if you comment [this line in YCM](https://github.com/Valloric/YouCompleteMe/blob/e018777b38eedaa23b96cfee40382d000e464e31/python/ycm/youcompleteme.py#L126)? It's starting a thread that keep the server alive. That thread is always running so it could potentially be the source of the crash. I don't think other threads are involved because they are short-lived and are only started on buffer change.

**Comment by dpelle** [comment]
My understanding is that signals are handled by one arbitrary thread of the process. 
If we don't want the Python process started by YouCompleteMe to handle the SIGCONT
signal (because it's handler is not thread safe), then that that should block SIGCONT,
see  https://docs.python.org/3.4/library/signal.html#signal.pthread_sigmask

If that's indeed what's causing the issue, then the bug is arguably not in Vim but
in the YouCompleteMe plugin.  Although, if the signal handler for SIGCONT (i.e.
sigcont_handler) was doing minimalistic actions, then it could probably work
whether Python thread(s) block this signal or not.  So making the Vim 
signal handler sigcont_handler() simpler would be preferable if possible.
I see that sigcont_handler() currently redraws the screen which does not
look right in a signal handler.

**Comment by dpelle** [comment]
@micbou asked:

> Does Vim still crash if you comment this line in YCM?

I made this change as suggested:
```
(master)~/.vim/bundle/YouCompleteMe$ git diff
diff --git a/python/ycm/youcompleteme.py b/python/ycm/youcompleteme.py
index 7231d12..bbed354 100644
--- a/python/ycm/youcompleteme.py
+++ b/python/ycm/youcompleteme.py
@@ -123,7 +123,7 @@ class YouCompleteMe( object ):
     self._server_is_ready_with_cache = False
     self._SetUpLogging()
     self._SetUpServer()
-    self._ycmd_keepalive.Start()
+    # self._ycmd_keepalive.Start()
 
 
   def _SetUpServer( self ):
```

But Vim still crashes when doing CTRL-Z followed by 'fg'  (not always, but often).
And tsan still always complains with many errors.  I do not see a difference.

If I disable YouCompleteMe by commenting out this line in my ~/.vimrc
then Vim never crashes:
```
"Plugin 'Valloric/YouCompleteMe'
```
So it tends to confirm that YouCompleteMe is triggering the issue.
But of course, this is not a solution since I'd like to use the
YouCompleteMe plugin.

**Comment by brammool** [comment]
Dominique wrote:

> > I cannot reproduce it, but I can guess a solution. I'll make a
> > patch, please try it out.
> 
> I tried to reproduce it with vim-8.1.301.
> Unfortunately, it still crashes once in a while when doing ":suspend" 
> followed by fg to get back to vim.  Sometimes, it does not crash
> immediately but Vim screen gets garbled. Sometimes it crashes
> immediately.  I see errors with both asan and tsan.  With asan, 
> the reported stacks are not always the same and tsan reports
> race conditions, which might explain why behavior is not always
> the same.

OK, it was worth a try.

Not sure why exactly the crash happens, I guess it's still some kind of
memory corruption.  I have no clue how to fix this.

I think the only sane thing to do now is to undo patch 8.1.0244.  That
means on receiving a SIGCONT Vim may be in a bad state, but hardly
anyone has complained about that.  Crashing is surely much worse.

> So according to tsan, the in_mch_suspend variable is accessed from 2
> different threads and it is not atomic nor protected by a mutex (hence
> race condition).  It cannot be protected by a mutex since it's used in
> signal handler.  Atomic variable are only availble in C11 I believe.
> 
> The fact that that there is another thread than the main thread seems to
>  be caused by the YouCompleteMe plugin that may spawn a thread, which 
> may not be safe given that Vim code is not thread safe in general. So
> perhaps @micbou or @Valloric have ideas about this bug.
> 
> I'll try to minimize my setup to reproduce it in a easier way.  But in a
> nutshell, reproducing it requires the YouCompleteMe plugin, then
> opening a file in Vim and doing  CTRL-Z (or ":suspend")  followed
> by the fg command in shell to get back to vim  (crash then sometimes
> happens). Crash happens with a without a sanitized build, but it's
> best to try to reproduce with a asan build (more chances to crashes
> and it gives better information when bug happens).

While rolling back the change that caused it, perhaps a solution is to
block signals in any thread but the main one.

But updating the screen in a signal handler probably is just asking for
trouble.  I also don't know a good way of having the signal handler
interrupt the normal processing (way is likely waiting for a character
to be typed) so that the screen updating is done outside of the signal
handler. Maybe setting a flag that tells the code to redraw helps, but
if Vim is blocked on reading input that flag won't be checked.

-- 
Computers make very fast, very accurate, mistakes.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by dpelle** [comment]
I just found why it happens.

Adding some printf(), I noticed that  sigcont_handler() was
called sometimes with in_mch_suspend being false whereas
it is meant to be true when doing CTRL-Z or :suspend.
This happens because mch_suspend does this:
1) call kill(0, SIGSTP)
2) set in_mch_suspend to false

But the sigcont_handler() may not be invoked immediately
and so 2) can be executed before sigcont_handler() is
invoked, causing trouble.

This bug was apparently already noticed for Solaris
a while ago since there was this comment in mch_suspend
but this code is #ifdef'ed out for me:
```
    /*
     * Wait for the SIGCONT signal to be handled. It generally happens
     * immediately, but somehow not all the time. Do not call pause()
     * because there would be race condition which would hang Vim if
     * signal happened in between the test of sigcont_received and the
     * call to pause(). If signal is not yet received, call sleep(0)
     * to just yield CPU. Signal should then be received. If somehow
     * it's still not received, sleep 1, 2, 3 ms. Don't bother waiting
     * further if signal is not received after 1+2+3+4 ms (not expected
     * to happen).
```

So to fix, it I removed the #ifdef _REENTRANT as below.  I tried
CTRL-Z  then fg many times and can no longer make it crash. 
Admittedly, tsan still complains about in_mch_suspend being
accessed in 2 threads but in practice it does not seem to cause
issues. I don't think it can be fixed without using C11 which 
introduced atomic variables.

I also changed some variables from int to sig_atomic_t but
that's not the main fix. The main fixing being to remove 
#ifdef _REETRANT. 

So please consider this patch that works for me on 
top of Vim-8.1.301:

[fix-crash-CTRL-Z-fg.patch.gz](https://github.com/vim/vim/files/2303837/fix-crash-CTRL-Z-fg.patch.gz)

```
diff --git a/src/os_unix.c b/src/os_unix.c
index b044df2..e7857b6 100644
--- a/src/os_unix.c
+++ b/src/os_unix.c
@@ -161,7 +161,7 @@ static int get_x11_title(int);
 static int get_x11_icon(int);
 
 static char_u	*oldtitle = NULL;
-static volatile int oldtitle_outdated = FALSE;
+static volatile sig_atomic_t oldtitle_outdated = FALSE;
 static int	did_set_title = FALSE;
 static char_u	*oldicon = NULL;
 static int	did_set_icon = FALSE;
@@ -205,7 +205,7 @@ static RETSIGTYPE catch_sigpwr SIGPROTOARG;
 # define SET_SIG_ALARM
 static RETSIGTYPE sig_alarm SIGPROTOARG;
 /* volatile because it is used in signal handler sig_alarm(). */
-static volatile int sig_alarm_called;
+static volatile sig_atomic_t sig_alarm_called;
 #endif
 static RETSIGTYPE deathtrap SIGPROTOARG;
 
@@ -231,13 +231,13 @@ static int save_patterns(int num_pat, char_u **pat, int *num_file, char_u ***fil
 #endif
 
 /* volatile because it is used in signal handler sig_winch(). */
-static volatile int do_resize = FALSE;
+static volatile sig_atomic_t do_resize = FALSE;
 static char_u	*extra_shell_arg = NULL;
 static int	show_shell_mess = TRUE;
 /* volatile because it is used in signal handler deathtrap(). */
-static volatile int deadly_signal = 0;	    /* The signal we caught */
+static volatile sig_atomic_t deadly_signal = 0;	    /* The signal we caught */
 /* volatile because it is used in signal handler deathtrap(). */
-static volatile int in_mch_delay = FALSE;    /* sleeping in mch_delay() */
+static volatile sig_atomic_t in_mch_delay = FALSE;    /* sleeping in mch_delay() */
 
 #if defined(FEAT_JOB_CHANNEL) && !defined(USE_SYSTEM)
 static int dont_check_job_ended = 0;
@@ -1243,9 +1243,9 @@ after_sigcont(void)
 
 #if defined(SIGCONT)
 static RETSIGTYPE sigcont_handler SIGPROTOARG;
-static int in_mch_suspend = FALSE;
+static volatile sig_atomic_t in_mch_suspend = FALSE;
 
-# if defined(_REENTRANT) && defined(SIGCONT)
+# if defined(SIGCONT)
 /*
  * On Solaris with multi-threading, suspending might not work immediately.
  * Catch the SIGCONT signal, which will be used as an indication whether the
@@ -1256,7 +1256,7 @@ static int in_mch_suspend = FALSE;
  *
  * volatile because it is used in signal handler sigcont_handler().
  */
-static volatile int sigcont_received;
+static volatile sig_atomic_t sigcont_received;
 # endif
 
 /*
@@ -1267,7 +1267,7 @@ sigcont_handler SIGDEFARG(sigarg)
 {
     if (in_mch_suspend)
     {
-# if defined(_REENTRANT) && defined(SIGCONT)
+# if defined(SIGCONT)
 	sigcont_received = TRUE;
 # endif
     }
@@ -1389,11 +1389,11 @@ mch_suspend(void)
     loose_clipboard();
 # endif
 
-# if defined(_REENTRANT) && defined(SIGCONT)
+# if defined(SIGCONT)
     sigcont_received = FALSE;
 # endif
     kill(0, SIGTSTP);	    /* send ourselves a STOP signal */
-# if defined(_REENTRANT) && defined(SIGCONT)
+# if defined(SIGCONT)
     /*
      * Wait for the SIGCONT signal to be handled. It generally happens
      * immediately, but somehow not all the time. Do not call pause()
```

**Comment by brammool** [comment]
Dominique wrote:

> I just found why it happens and I will create a PR with a fix.
> 
> Adding some printf(), I noticed that  sigcont_handler() was
> called sometimes with in_mch_suspend being false whereas
> it is meant to be true when doing CTRL-Z or :suspend.
> This happens because mch_suspend does this:
> 1) call kill(0, SIGSTP)
> 2) set mch_suspend to false
> 
> But the sigcont_handler() may not be invoked immediately
> and so 2) can be executed before sigcont_handler() is
> invoked, causing trouble.
> 
> This bug was apparently already noticed for Solaris
> a while ago since there was this comment in mch_suspend
> but this code is #ifdef'ed out for me:
> ```
>     /*
>      * Wait for the SIGCONT signal to be handled. It generally happens
>      * immediately, but somehow not all the time. Do not call pause()
>      * because there would be race condition which would hang Vim if
>      * signal happened in between the test of sigcont_received and the
>      * call to pause(). If signal is not yet received, call sleep(0)
>      * to just yield CPU. Signal should then be received. If somehow
>      * it's still not received, sleep 1, 2, 3 ms. Don't bother waiting
>      * further if signal is not received after 1+2+3+4 ms (not expected
>      * to happen).
> ```
> 
> So to fix, it I removed the #ifdef _REENTRANT as below.  I tried
> CTRL-Z  then fg many times and can no longer make it crash. 
> Admittedly, tsan still complains about in_mch_suspend being
> accessed in 2 threads but in practice it does not seem to cause
> issues. I don't think it can be fixed without using C11 which 
> introduced atomic variables.
> 
> I also changed some variables from int to sig_atomic_t but
> that's not the main fix. The main fixing being to remove 
> #ifdef _REETRANT. 
> 
> So please consider this patch that works for me on 
> top of Vim-8.1.301:

I'm very glad you found why this happens.  Makes a lot of sense.

Still, calling screen_update() in the signal handler should probably be
avoided.  The signal stack might be small and with syntax highlighting a
lot of work might be done.  But we can at least put the terminal back in
raw mode, so that most things work.

-- 
Kiss me twice.  I'm schizophrenic.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by vim-ml** [comment]
On Mon, Aug 20, 2018 at 12:25:33PM -0700, Bram Moolenaar wrote:
> But updating the screen in a signal handler probably is just asking for
> trouble. I also don't know a good way of having the signal handler
> interrupt the normal processing (way is likely waiting for a character
> to be typed) so that the screen updating is done outside of the signal
> handler. Maybe setting a flag that tells the code to redraw helps, but
> if Vim is blocked on reading input that flag won't be checked.

In theory, if Vim receives a signal while it's blocked on reading, the
read syscall will fail with EINTR.  At that point any flags set by the
signal handler can be checked and the screen can be redrawn before
retrying the read to wait for more input.

**Comment by chrisbra** [comment]
I suppose this is fixed by f1883479be91550bc3, 2e31048c304fb94d62556 and 8e82c057ffb86cec so closing.
